Removed hard-coded dependency on yokozuna

fadushin opened the pull request:
This PR removed a hard-coded dependency on yokozuna by indirecting calls into yokozuna through a module that is loaded via configuration, using the new `index_module` (internally defined) configuration variable.

By default (for the purposes of riak_kv, by itself), the `index_module` configuration variable is undefined, which will result in a no-op during puts, deletes, and handoff operations from the vnode.

Downstream repos (notably, yokozuna), configure `index_module` to be a module the `riak_kv_vnode` will call as part of the above set of operations.

The `index_module` configuration variable is not intended to be shared with users, as its only function is to indirect calls into an indexing subsystem.

Dependent PRs: -none-
Downstream PRs: https://github.com/basho/yokozuna/pull/716

## Checklist

- [x] Observability
This PR has no impact on a running system, at least as we deploy riak today.  Any new configuration we add is hidden and should get picked up automatically as part of an upgrade.
- [x] Performance
In general this change should have no impact on performance.  In the case where yokozuna is not deployed with riak_kv (no known deployments), this change should eliminate calls into yokozuna, theoretically resulting in fewer reductions
- [x] Fault-tolerance
There are no remote processes or IPC added as part of this PR
- [x] Testing
This PR does not affect code paths in any existing tests.  No specific tests have been added for this PR, as there is no new functionality being added.
- [x] Code Clarity/Quality
- [x] Code Documentation
The rationale for the index_module configuration documentation is explained in the source code documentation.
- [x] External Documentation
n/a
- [x] Release Notes
n/a

fadushin committed:
Removed had-coded dependency on yokozuna, and instead call indexing functions indirectly through a module defined in configuration (`undefined`, by default, which results in a no-op).

JeetKunDoug commented on src/riak_kv_vnode.erl:
```diff
@@ -92,19 +92,6 @@
--endif.
-
 -ifdef(TEST).
 -define(YZ_SHOULD_HANDOFF(X), true).
```
This is still calling `yz_kv` directly in the non-test case - of course Github won't let me comment on _that_ line, but this is close enough.

jvoegele commented on src/riak_kv_vnode.erl:
```diff
@@ -2704,6 +2697,45 @@ highest_actor(ActorBase, Obj) ->
+%%     - index(object_pair(), write_reason(), p()) -> ok.
+%%     - index_binary(bucket(), key(), binary(), write_reason(), p()) -> ok.
+%%     - is_searchable(riak_kv_bucket:props()) -> boolean().
+%%
```
Would it be feasible to define these functions as callbacks in a behavior module? If so, the `yokozuna.erl` module (and any other future indexing modules that we might create) could use the behavior to get some extra compile-time checking.

JeetKunDoug replied:
+1 - good idea.

JeetKunDoug commented on src/riak_kv_vnode.erl:
```diff
@@ -2704,6 +2697,45 @@ highest_actor(ActorBase, Obj) ->
+%%
+%%     - index(object_pair(), write_reason(), p()) -> ok.
+%%     - index_binary(bucket(), key(), binary(), write_reason(), p()) -> ok.
+%%     - is_searchable(riak_kv_bucket:props()) -> boolean().
```
As per above - need to add `should_handoff` with appropriate stuff.

JeetKunDoug replied:
Still need to actually add `should_handoff` here, or maybe just remove the function list here and refer to the behaviour now that we have it?

fadushin committed:
Defined an update module behavior, and tweaked names per PR comments and off-line discussions.  Added some type specs to help dialyzer make sense of the world.

JeetKunDoug commented on src/riak_kv_vnode.erl:
```diff
@@ -134,6 +115,8 @@
           leasing = false :: boolean()
          }).
 
+-type update_hook() :: module() | undefined.
```
Thoughts on defaulting to a `riak_kv_noop_update_hook` module, rather than `undefined` here? Implement the no-op versions of all the functions and we can get rid of all of the function heads that check for `undefined` and just call the noop version... and change all the `maybe_*` functions to just `*`.

jvoegele replied:
Seems like a win to me.

JeetKunDoug reviewed (commented):
2 suggestions below, but otherwise +1 - I like it.

fadushin committed:
Fixed comment to accurately reflect reality.

Commit:
Make a "no-op" update hook

Creating a default, no-op update hook allows us to get rid of all the
special-casing of `undefined` when dealing with update hooks - there will always be one available, and the default will just do nothing.

Commit:
Add parameter names to callback definition of `riak_kv_update_hook`.

While the types are quite descriptive, this helps autogenerate
code in editors that can understand the names as well as types.

hazen commented:
Just excited to see modules decoupling!

fadushin committed:
Merge branch 'develop' into fd-remove-yz-dep

fadushin committed:
Added hash tree action type to make type spec correct.

nickelization commented:
+1
